### PR TITLE
Add upgrade procedure for Channels DB v8 to v10

### PR DESCRIPTION
**Pull request by ichorid**
The upgrade procedure creates a temporary Channels DB and directly dumps the old Channels tables into the corresponding tables in the new DB.
If the upgrade finishes correctly, the new DB replaces the old one.
If the upgrade is skipped by the user, the old DB is removed
If the upgrade crashes, at the next Tribler startup it will be tried anew.

The main reason to upgrade is the removal of unused unique indexes that mess up the new DB logic. Unfortunately, it is impossible to remove them without recreating the whole database.

5 GB database converts in 5 minutes on my PC.

### CODE REVIEW

**Comment by ichorid** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

**Comment by devos50** [comment]
retest this please

File: src/tribler-core/tribler_core/upgrade/db8_to_db10.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
+import contextlib
+import datetime
+import logging
+import sqlite3
+from asyncio import sleep
+
+TABLE_NAMES = (
+    "ChannelNode", "TorrentState", "TorrentState_TrackerState", "ChannelPeer", "ChannelVote", "TrackerState", "Vsids")
```
Comment (kozlovsky) [inline src/tribler-core/tribler_core/upgrade/db8_to_db10.py:8]: Should the `MiscData` table also be on this list?

Comment (ichorid) [reply src/tribler-core/tribler_core/upgrade/db8_to_db10.py:8]: No. MiscData is currently only used for upgrader marks and database version.

**Comment by ichorid** [comment]
retest this please

File: src/tribler-core/tribler_core/upgrade/db8_to_db10.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    def get_table_entries_count(self, cursor, table_name):
+        return cursor.execute(f"SELECT COUNT(*) FROM {table_name};").fetchone()[0]
+
+    async def convert_table(self, cursor, table_name, column_names):
+        def convert_command(offset, batch_size):
+            sql_command = None
+            try:
+                sql_command = f"INSERT OR IGNORE INTO {table_name} ({column_names}) " + \
+                              f"SELECT {column_names} FROM old_db.{table_name} " + \
+                              f"LIMIT {batch_size} {('OFFSET ' + str(offset)) if offset else ''} ;"
```
Comment (kozlovsky) [inline src/tribler-core/tribler_core/upgrade/db8_to_db10.py:73]: I'd add `ORDER BY  ROWID` just to be sure that SELECT returns rows in a stable order

**Comment by kozlovsky** [comment]
Screenshot 2020-11-30 at 19 45 56

Currently, when I press the "Skip" button during the upgrade I see the window with the following message: "Abort the conversion of old channels. The upgrade procedure is now converting torrents in channels collected by the previous installation of Tribler. Are you sure you want to abort the conversion process?".

If I understand correctly, choosing "Abort" will delete the previous database without the possibility to recover it. The message does not warn about it. It can be understood as "the upgrade is optional, you can abort conversion procedure and then continue to use your old database". We need to warn the user that all data will be lost if pressing the "Abort" button.

**Comment by ichorid** [comment]
изображение


Is it better now? :wink:

**Comment by kozlovsky** [comment]
image

When I started the upgrade the speed of copying metadata info was about 1000 rows/seconds, and now it is about 200 rows/seconds. It seems that all upgrade works as a single transaction, and big transactions are not always efficient. Probably it should be split into several transactions of smaller size.

**Comment by kozlovsky** [comment]
> Is it better now? 😉

Definitely. But the user will lose not only the personal channel but all other loaded channels as well.

**Review by kozlovsky — APPROVED** [review]

**Review by kozlovsky — APPROVED** [review]

**Pull request merged**